### PR TITLE
VSR: Fix allocation size of parse_addresses() return value

### PR DESCRIPTION
**Pull request by ifreund**
The Allocator interface requires the exact length of the allocation to be provided when freeing memory. With the current implementation, there is no correct way to free the returned slice.

### CODE REVIEW

**Review by jorangreef — COMMENTED** [review]
Thanks @ifreund good catch!

`shrink()` like this works.

However, we've added `StaticAllocator` to Replica which has a principle of only allowing allocation, and no interim frees etc. We want to see how far we can push this, to force allocation sites to always allocate exactly, without "fixer uppers".

Therefore, would you please change this to use something like `mem.count()` on the number of commas, so that we allocate exactly without the shrink?

**Comment by ifreund** [comment]
> Therefore, would you please change this to use something like mem.count() on the number of commas, so that we allocate exactly without the shrink?

Done, StaticAllocator seems like a neat idea :)

File: src/vsr.zig [hunk 1]
```diff
@@ -815,18 +815,22 @@ test "exponential_backoff_with_jitter" {
 /// This does require that the user specify the same order to all replicas.
 /// The caller owns the memory of the returned slice of addresses.
 pub fn parse_addresses(allocator: std.mem.Allocator, raw: []const u8, address_limit: usize) ![]std.net.Address {
-    // TODO After parsing addresses resize the memory allocation.
-    var addresses = try allocator.alloc(std.net.Address, address_limit);
+    const address_count = std.mem.count(u8, raw, ",") + 1;
```
Comment (batiati) [inline src/vsr.zig:818]: We already count the commas at [c/tb_client/thread.zig](https://github.com/tigerbeetledb/tigerbeetle/blob/a15fa7523385b826d85c6294f374314369f3913a/src/c/tb_client/thread.zig#L101). Would it be better to remove the `address_limit` parameter and use `config.replicas_max` instead?

Comment (jorangreef) [reply src/vsr.zig:818]: Thanks @batiati 

I think that `c/tb_client/thread.zig` shouldn't count, and delegate this rather to `parse_addresses()`, passing `config.replicas_max` as `address_limit`. `address_limit` is useful as a parameter here to decouple from `config`, i.e. for testing.

**Review by jorangreef — DISMISSED** [review]
Thanks @ifreund this looks good.

Do the tests need to be updated at all for this? Is there any coverage that could be added?

**Comment by ifreund** [comment]
> Do the tests need to be updated at all for this? Is there any coverage that could be added?

I had to tweak one test due to a slight change in the order of errors reported since we now check the limit before checking that the addresses are valid. 

I also just added one positive test case that explicitly tests when the number of provided addresses is less than the limit. I don't see any other missing edge cases in the tests.

**Review by jorangreef — APPROVED** [review]
